Panic when downloading pprof

Sadly when downloading pprof profiles Parca can panic.

```
name=parca ts=2021-11-04T18:08:35.221310105Z caller=log.go:168 level=info msg="loading bucket configuration"
level=info name=parca ts=2021-11-04T18:08:35.221566928Z caller=server.go:85 msg="starting server" addr=:7070
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x27f5787]

goroutine 43124 [running]:
github.com/parca-dev/parca/pkg/storage.GeneratePprof.func1(0x3994aa0, 0xc02d063080, 0xc02d063080, 0x0)
	github.com/parca-dev/parca/pkg/storage/pprof.go:147 +0x2a7
github.com/parca-dev/parca/pkg/storage.WalkProfileTree(0x3972b68, 0xc02d058648, 0xc02d04efb0, 0xc00294c3d0, 0xb)
	github.com/parca-dev/parca/pkg/storage/interface.go:63 +0xcc
github.com/parca-dev/parca/pkg/storage.GeneratePprof(0x39a88f0, 0xc02d017590, 0x7f3bae849788, 0xc0005c01d0, 0x3972aa0, 0xc02d06b9a0, 0xc00031ce80, 0x39a88f0, 0xc02d017860)
	github.com/parca-dev/parca/pkg/storage/pprof.go:82 +0x431
github.com/parca-dev/parca/pkg/query.(*Query).renderReport(0xc00031ce80, 0x39a88f0, 0xc02d017590, 0x3972aa0, 0xc02d06b9a0, 0xc000000001, 0x0, 0x0, 0x1)
	github.com/parca-dev/parca/pkg/query/query.go:297 +0xbf
github.com/parca-dev/parca/pkg/query.(*Query).singleRequest(0xc00031ce80, 0x39a88f0, 0xc02d017590, 0xc02d035380, 0xc000000001, 0xc02c6594b0, 0xc02c659468, 0x40bc25)
	github.com/parca-dev/parca/pkg/query/query.go:201 +0xc5
github.com/parca-dev/parca/pkg/query.(*Query).Query(0xc00031ce80, 0x39a88f0, 0xc02d017590, 0xc02d02b130, 0xc00031ce80, 0x2bc2f80, 0xc02d040c30)
	github.com/parca-dev/parca/pkg/query/query.go:166 +0x9d
github.com/parca-dev/parca/gen/proto/go/parca/query/v1alpha1._QueryService_Query_Handler.func1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0x0, 0x4d43, 0x5, 0x33d7124)
	github.com/parca-dev/parca/gen/proto/go/parca/query/v1alpha1/query_grpc.pb.go:164 +0x89
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors.UnaryServerInterceptor.func1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0xc02d01f640, 0xc02d0322e8, 0x3, 0x3, 0x39c2318, 0xc0006793e0)
	github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20201207153454-9f6bf00c00a7/interceptors/server.go:22 +0x239
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0x28, 0xc02d02b220, 0x3379c3e, 0x8)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0xc02d01f640, 0xc02d01f660, 0x0, 0x0, 0x0, 0x0)
	github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xb0
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0x3379c3e, 0x8, 0x0, 0x0)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1(0x39a88f0, 0xc02d017590, 0x317bde0, 0xc02d02b130, 0xc02d01f640, 0xc02d01f680, 0x0, 0x0, 0x0, 0x0)
	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.24.0/interceptor.go:325 +0x6b6
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x39a88f0, 0xc02d0174a0, 0x317bde0, 0xc02d02b130, 0xc00006e400, 0x0, 0xc02c659b30, 0x40e3b8)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x39a88f0, 0xc02d0174a0, 0x317bde0, 0xc02d02b130, 0xc02d01f640, 0xc02d0322e8, 0xc004e5aba0, 0x4a2d66, 0x30ea8e0, 0xc02d0174a0)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34 +0xd7
github.com/parca-dev/parca/gen/proto/go/parca/query/v1alpha1._QueryService_Query_Handler(0x31dea80, 0xc00031ce80, 0x39a88f0, 0xc02d0174a0, 0xc02d03ba40, 0xc00056c060, 0x39a88f0, 0xc02d0174a0, 0xc02aa4c900, 0x111)
	github.com/parca-dev/parca/gen/proto/go/parca/query/v1alpha1/query_grpc.pb.go:166 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000401340, 0x39cf078, 0xc02d02c7e0, 0xc02cc17200, 0xc00056c300, 0x4f6e218, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.41.0/server.go:1279 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000401340, 0x39cf078, 0xc02d02c7e0, 0xc02cc17200, 0x0)
	google.golang.org/grpc@v1.41.0/server.go:1608 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc02d0363a0, 0xc000401340, 0x39cf078, 0xc02d02c7e0, 0xc02cc17200)
	google.golang.org/grpc@v1.41.0/server.go:923 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.41.0/server.go:921 +0x1fd
```

--- a/README.md
+++ b/README.md
@@ -74,4 +74,4 @@ Parca was originally developed by [Polar Signals](https://polarsignals.com/). Re
 ## Contributing
 
 Check out our [Contributing Guide](CONTRIBUTING.md) to get started!
-
+It explains how compile Parca, run it with Tilt as container in Kubernetes and send a Pull Request.

--- a/pkg/storage/pprof.go
+++ b/pkg/storage/pprof.go
@@ -140,11 +140,16 @@ func GeneratePprof(ctx context.Context, locationStore Locations, ip InstantProfi
 				}
 			}
 
-			location := &profile.Location{
-				ID:      uint64(len(p.Location) + 1),
-				Mapping: mapping,
+			addr := loc.Address
+			if mapping != nil {
 				// TODO: Is this right?
-				Address:  loc.Address + mapping.Offset,
+				addr += mapping.Offset
+			}
+
+			location := &profile.Location{
+				ID:       uint64(len(p.Location) + 1),
+				Mapping:  mapping,
+				Address:  addr,
 				Line:     lines,
 				IsFolded: loc.IsFolded,
 			}

--- a/pkg/storage/pprof_test.go
+++ b/pkg/storage/pprof_test.go
@@ -21,6 +21,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/google/pprof/profile"
+	"github.com/google/uuid"
 	"github.com/parca-dev/parca/pkg/storage/metastore"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/require"
@@ -57,6 +58,53 @@ func TestGeneratePprof(t *testing.T) {
 	require.NoError(t, tmpfile.Close())
 
 	f, err = os.Open(tmpfile.Name())
+	require.NoError(t, err)
+	resProf, err := profile.Parse(f)
+	require.NoError(t, err)
+	require.NoError(t, f.Close())
+	require.NoError(t, resProf.CheckValid())
+}
+
+func TestGeneratePprofNilMapping(t *testing.T) {
+	ctx := context.Background()
+
+	pt := NewProfileTree()
+	pt.Insert(makeSample(2, []uuid.UUID{
+		uuid.MustParse("00000000-0000-0000-0000-000000000002"),
+		uuid.MustParse("00000000-0000-0000-0000-000000000001"),
+	}))
+
+	l := &fakeLocations{m: map[uuid.UUID]*metastore.Location{
+		uuid.MustParse("00000000-0000-0000-0000-000000000001"): {
+			ID: uuid.MustParse("00000000-0000-0000-0000-000000000001"),
+			Lines: []metastore.LocationLine{{
+				Function: &metastore.Function{
+					ID:          uuid.MustParse("00000000-0000-0000-0000-0000000000f1"),
+					FunctionKey: metastore.FunctionKey{Name: "1"},
+				},
+			}},
+		},
+		uuid.MustParse("00000000-0000-0000-0000-000000000002"): {
+			ID: uuid.MustParse("00000000-0000-0000-0000-000000000002"),
+			Lines: []metastore.LocationLine{{
+				Function: &metastore.Function{
+					ID:          uuid.MustParse("00000000-0000-0000-0000-0000000000f2"),
+					FunctionKey: metastore.FunctionKey{Name: "2"},
+				},
+			}},
+		},
+	}}
+
+	res, err := GeneratePprof(ctx, l, &Profile{Tree: pt})
+	require.NoError(t, err)
+
+	tmpfile, err := ioutil.TempFile("", "pprof")
+	defer os.Remove(tmpfile.Name())
+	require.NoError(t, err)
+	require.NoError(t, res.Write(tmpfile))
+	require.NoError(t, tmpfile.Close())
+
+	f, err := os.Open(tmpfile.Name())
 	require.NoError(t, err)
 	resProf, err := profile.Parse(f)
 	require.NoError(t, err)

